### ISSUE TITLE
failsafe vertex labels to examing shortening

### ISSUE THREAD

**Issue by yakra** [TravelMapping/DataProcessing#90]
These can be copied to the comments of siteupdate.py.

`A9Gun/A9Alt/E49Gun/E49Alt/E51Gun/E51Alt@ST/SN`

**Comment by yakra** [comment]
US6@BriSt&MA79@*BriSt&MA138@BriSt

**Comment by yakra** [comment]
US54BusElP@US54_E&US54ElP@US54Bus_E

**Comment by yakra** [comment]
ME24BusBru@US201_N&US201@ME24Bus_N

**Comment by yakra** [comment]
```
3+ intersection: US16@US16Trk/385Trk_S&US16TrkHil@US16/385_S&US385@US16Trk/385Trk_S&US385TrkHil@US16/385_S	US16/US16TrkHil/US385/US385TrkHil
3+ intersection: US16@US16Trk/385Trk_N&US16TrkHil@US16/385_N&US385@US16Trk/385Trk_N&US385TrkHil@US16/385_N	US16/US16TrkHil/US385/US385TrkHil
Appended region: US16/US16TrkHil/US385/US385TrkHil|SD
```

**Comment by jteresco** [comment]
The 100 longest labels as of the most recent update are below, which you can upload to HDX, if you save as a .wpl file, to see where they are.  Business routes kill the current waypoint simplification rules.

<pre>US129BusWat@US129_N&US129@US129/441BusWat_N&US441BusWat@US441_N&US441@US129/441BusWat_N&GA15@US129/441&GA24@US129/441BusWat_N&GA24BusWat@GA24_N 33.880608 -83.41524
US29@GA10Lp(8)&US78BusAth@US78_E&US78@GA10Lp(8)&US129@GA10Lp(8)&US441@GA10Lp(8)&GA8@GA10Lp(8)&GA10@GA10Lp(8)&GA10LpAth@8&GA15@GA10Lp(8) 33.946733 -83.356115
US129BusEat@US129_S&US129@US129/441Bus_S&US441BusEat@US441_S&US441@US129/441Bus_S&GA24@US129/441Bus_S&GA24BusEat@GA24_S&GA44@US129/441 33.297095 -83.384664
I-69FutInd@53(465)&I-65@106&I-74@53(465)&US31@53(465)&US36@I-465(53)&US40@I-465(53)&I-465@I-65&I-465@53&IN37@I-465(53)&IN67@I-465(53) 39.703423 -86.108265
US129BusWat@US129_S&US129@US129/441BusWat_S&US441BusWat@US441_S&US441@US129/441BusWat_S&GA24@US129/441BusWat_S&GA24BusWat@GA24_S 33.845643 -83.42296
AH5Shy@KAZ/UZB&AH5@KAZ/UZB&AH62@KAZ/UZB&AH62@KAZ/UZB&M39@KAZ/UZB&E40@UZB/KAZ&E40Shy@UZB/KAZ&E123@KAZ/UZB&E123@KAZ/UZB&A2@KAZ/UZB 41.471448 69.358535
I-25@186&I-25BLCas@I-25_S&US20BusCas@US20_E&US20@I-25(186)&US26BusCas@US26_E&US26@I-25(186)&US87BusCas@US87_S&US87@I-25(186) 42.85662 -106.293969
I-25@140&I-25BLDou@I-25_N&US20BusDou@US20_W&US20@I-25(140)&US26BusDou@US26_W&US26@I-25(140)&US87BusDou@US87_N&US87@I-25(140) 42.759333 -105.412552
I-25@135&I-25BLDou@I-25_S&US20BusDou@US20_E&US20@I-25(135)&US26BusDou@US26_E&US26@I-25(135)&US87BusDou@US87_S&US87@I-25(135) 42.732171 -105.341506
I-69FutInd@49(465)&I-74@94&US31@49(465)&US36@I-465(49)&US40@I-465(49)&I-465@49&IN37@I-465(49)&IN67@I-465(49)&US421@I-465(49) 39.732043 -86.04475
I-69FutInd@44(465)&I-70@90&US31@44(465)&US36@I-465(44)&I-465@44&US52@I-465(44)&IN37@I-465(44)&IN67@I-465(44)&US421@I-465(44) 39.798942 -86.034837
US521BusKer@US601/601Bus_N&US601BusKer@US601_N&US601@US521/601Bus_N&SC157TrkKer@US521Bus/601Bus&SC341TrkKer@US521Bus/601Bus 34.551819 -80.583811
US19TrkPit@I-376(69B)_S&US19TrkPit@I-376(69B)_N&US19@I-376(69B)&US22@I-376(69B)&US30@I-376(69B)&I-376@69B&PA51@I-376(69B) 40.430926 -80.026302
US29@GA10Lp(6)&US78@GA10Lp(6)&US129@GA10Lp(6)&US441@GA10Lp(6)&GA8@GA10Lp(6)&GA10LpAth@6&GA15@GA10Lp(6)&GA15AltAth@GA15_S 33.921141 -83.378216
US129BusEat@US129_N&US129@US129/441Bus_N&US441BusEat@US441_N&US441@US129/441Bus_N&GA24@US129/441Bus_N&GA24BusEat@GA24_N 33.347182 -83.392823
US129BypMad@US129_S&US129@US129/441Byp_S&US441BypMad@US441_S&US441@US129/441Byp_S&GA24@US129/441Byp_S&GA24BypMad@GA24_S 33.564552 -83.47831
US15BusWar@US15_S&US15@US15/29BusWar_S&US17BusWar@US17_S&US17@US15/29BusWar_S&US29BusWar@US29_S&US29@US15/29BusWar_S 38.686448 -77.787194
US19BusAlb@US19_S&US19@US19/82Bus&US82BusAlb@US19&GA3@US19/82Bus&GA133@US19/82Bus&GA300@US19/82Bus&GA520BusAlb@US19 31.571644 -84.111462
US129BusHaw@US129_N&US129@US129/341Bus&US341BusHaw@US341_N&US341@US129/341Bus&GA11@US129/341Bus&GA11BusHaw@GA11_N 32.292905 -83.482425
AH5@KAZ/KGZ&AH5@KAZ/KGZ&AH61@KGZ/KAZ&AH61@KGZ/KAZ&E40@KGZ/KAZ&E40@KGZ/KAZ&E125@KAZ/KGZ&E125@KAZ/KGZ&EM01@KGZ/KAZ 43.025388 74.704885
I-24Cha@178A&US11@I-24(178A)&US27@I-24(178A)&US41@I-24(178A)&US64@I-24(178A)&US72Cha@I-24(178A)&TN17@I-24(178A) 35.030778 -85.314964
US23BusGat@US23_N&US23@US23/58Bus_W&US58BusGat@US58_W&US58@US23/58Bus_W&US421BusGat@US421_N&US421@US23/58Bus_W 36.632181 -82.611469
US23BusGat@US23_S&US23@US23/58Bus_E&US58BusGat@US58_E&US58@US23/58Bus_E&US421BusGat@US421_S&US421@US23/58Bus_E 36.637759 -82.566816
US13BusSuf@US13_N&US13@US13/58Bus_E&US58BusSuf@US58_E&US58@US13/58Bus_E&US460BusSuf@US460_E&US460@US13/58Bus_E 36.752493 -76.523563
I-55@TN/AR&I-55@TN/AR&US61@TN/AR&US61@TN/AR&US64@AR/TN&US64@AR/TN&US70@AR/TN&US70@AR/TN&US79@AR/TN&US79@AR/TN 35.128262 -90.076736
I-26@3A(240)&US19BusAsh@US19_N&US19@I-240(3A)&US23BusAsh@US23_N&US23@I-240(3A)&I-240@3A&US74AltAsh@I-240(3A) 35.589167 -82.57998
US25BypSta@US25_S&US25@US25/301Byp&US301BypSta@US301_S&US301@US25/301Byp&GA67BypSta@US25_S&GA73@US25/301Byp 32.416333 -81.79965
US13BusSuf@US58/460Bus_W&US58BusSuf@US13/460Bus_S&US460BusSuf@US13/58Bus_S&VA10@US13/58Bus&VA32@US13/58Bus 36.737687 -76.582603
E5@FRA/ESP&E5@FRA/ESP&AP1@ESP/FRA&AP8@FRA/ESP&E70@ESP/FRA&E70@ESP/FRA&E80@ESP/FRA&E80@ESP/FRA&A63@ESP/FRA 43.339904 -1.752577
I-90BLShe@I-90_W&I-90@20&US14BusShe@US14_W&US14@I-90(20)&US87BusShe@US87_N&US87She@I-90(20)&WY338@I-90 44.838373 -106.972461
US29@GA10Lp(4)&US78@GA10Lp(4)&US129@GA10Lp(4)&US441@GA10Lp(4)&GA8@GA10Lp(4)&GA10LpAth@4&GA15@GA10Lp(4) 33.916325 -83.405167
US29@GA10Lp(7)&US78@GA10Lp(7)&US129@GA10Lp(7)&US441@GA10Lp(7)&GA8@GA10Lp(7)&GA10LpAth@7&GA15@GA10Lp(7) 33.933426 -83.367251
US14@US89/191&US16@US89/191&US89@US14/16&US191@US14/16&US287@US14/16&GraLpRd@SEntRd&SEntRd@GraLoopRd 44.414049 -110.578616
I-530@46&US65BusPin@US65_S&US63@I-530(46)&US65@I-530(46)&US79@I-530(46)&US425@I-530&AR190@I-530(46) 34.202929 -91.966896
US29@GA10Lp(10B)&US129@GA10Lp(10B)&US441@GA10Lp(10B)&GA8@GA10Lp(10B)&GA10LpAth@10B&GA15@GA10Lp(10B) 33.978056 -83.352413
US29@GA10Lp(10A)&US129@GA10Lp(10A)&US441@GA10Lp(10A)&GA8@GA10Lp(10A)&GA10LpAth@10A&GA15@GA10Lp(10A) 33.975885 -83.35443
US1@US521/601Trk_S&US521TrkCam@US1/601_S&US601TrkCam@US601_S&US601@US521/601Trk_S&SC34@US521/601Trk 34.249627 -80.645781
I-40BLWin@6A&US52@I-40BL&US158@I-40BL(6A)&US311@I-40BL&US421@I-40BL(6A)&NC8@I-40BL&NC150@I-40BL(6A) 36.094054 -80.232511
US20@MT/WY&US20@MT/WY&US191@WY/MT&US191WYe@WY/MT&US287@WY/MT&US287@WY/MT&WEntRd@MT/WY&WEntRd@MT/WY 44.651498 -111.055298
US1@US1/23Bus_N&US1BusWay@US1_N&US23BusWay@US23_N&US23@US1/23Bus_N&GA4@US1/23Bus_N&GA4BusWay@GA4_N 31.290885 -82.448283
US1@US1/78Trk_W&US1TrkAik@US1_S&US78TrkAik@US78_W&US78@US1/78Trk_W&SC19TrkAik@US1/78&SC118@US1/78 33.561647 -81.778418
US321TrkBoo@US321_N&US421TrkBoo@US421_N&US321@US321Trk/421Trk&US421@US321Trk/421Trk&NC194@NC105 36.230272 -81.720396
US1@US421/87Byp&US15@US421/87Byp&US421@149&US501@US421/87Byp&NC87@US421/87Byp&NC87BypSan@NC87_N 35.51119 -79.188037
I-69FutInd@37(465)&I-69@200&US31@37(465)&I-465@37&US52@I-465(37)&IN37@I-69/465&US421@I-465(37) 39.894855 -86.054149
E20@E47/E55_S&KogBugMot@Vest/SydMot&SydMot@KogeBugtMot&VestMot@KogeBugtMot&E47@E20_W&E55@E20_W 55.479486 12.137146
US221TrkBoo@US321/421Trk_N&US321TrkBoo@US221Trk_S&US421TrkBoo@US221Trk_S&NC105@US321/421Trk_N 36.206858 -81.704764
US11E@US11/19Trk&US11TrkBri@US11E/19&US19@US11/19Trk_S&US19TrkBri@US11E/19_S&US421@US11/19Trk 36.596115 -82.188742
I-30@141B&US70BusLit@US70_E&US65@I-30(141B)&US67@I-30(141B)&US70@I-30(141B)&US167@I-30(141B) 34.755978 -92.262363
I-25@188A&I-25BLCas@I-25_N&US20@I-25(188A)&US26@I-25(188A)&US87BusCas@US87_N&US87@I-25(188A) 42.857029 -106.325147
I-74FutWhi@242&US74BusWhi@US74_E&US76BusWhi@US76_E&US74@US74Bus/76Bus_E&US76@US74Bus/76Bus_E 34.349212 -78.658183
E261@UliWojPol&S5Swi@UliWojPol&DK5Wia@UliWojPol&S91@UliWojPol&DK91Gda@UliWojPol&DW239@S5/S91 53.419878 18.479948
I-40@237&I-40BLHen@I-40_W&US62BusHen@US62_W&US75BusHen@US75_S&US62@I-40(237)&US75@I-40(237) 35.433156 -96.016903
I-30@138&I-440@I-30/530&I-530@1&US65@I-30(138)&US67@I-30(138)&US70@I-30(138)&US167@I-530(1) 34.710856 -92.270994
US20@US89/191&US89@US20/191&US191@US20/89_N&US287@US20/89_N&GraLpRd@WEntRd&WEntRd@GraLoopRd 44.645391 -110.85852
US25BusRic@US25_S&US25@US25/421Bus_S&US421BusRic@US421_S&US421@US25/421Bus_S&KY876@US25/421 37.733572 -84.274288
US45WBusHum@US45W_N&US45W@US79/70AltByp&US70AltBypHum@US45W/45WBus&US79BypHum@US45W/45WBus 35.83722 -88.915089
US25BusGre@US25_N&US25@US25/178Bus_N&US178BusGre@US178_W&US178@US25/178Bus_N&SC72@US25/178 34.211396 -82.17778
US17BusWar@US211/211Bus_W&US29BusWar@US211/211Bus_W&US211BusWar@US211_W&US211@US17/29Bus_S 38.717268 -77.808073
AH5@KGZ/KAZ&AH5Shy@KGZ/KAZ&AH7@KAZ/KGZ&AH7@KAZ/KGZ&E40Shy@KAZ/KGZ&E40@KAZ/KGZ&EM01@KAZ/KGZ 42.823232 73.510766
I-80BLRaw@I-80_E&I-80@215&US30BusRaw@US30_E&US30@I-80(215)&US287@I-80(215)&WY76@I-80(215) 41.789137 -107.20047
I-72@157&US36BusHan@US36_E&US36@I-72(157)&US61BusHan@I-72&MO79@I-72/36&MO110Han@I-72(157) 39.71714 -91.372894
US19BusGri@US19_N&US19@US19/41Bus_N&US41BusGri@US41_N&US41@US19/41Bus_N&GA3@US19/41Bus_N 33.264146 -84.28977
AmaMot@MotRing3&E20@E47/E55_N&KogBugMot@MotRing3&MotRing3@Ama/OreMot&E47@E20_E&E55@E20_E 55.629922 12.429914
I-49@84&US167BusAle@US167_N&US167@I-49(84)&LA1@I-49(84)&LA28@I-49(84)&LA28BusAle@LA28_E 31.303351 -92.447129
I-16@164&US17@I-16(164)&I-516@5&US80@I-516(5)&GA21@I-516(5)&GA25@I-516(5)&GA26@I-516(5) 32.071511 -81.13501
I-87FutRoc@435(64)&US64BusWen@US64_E&US64@US64BusWen_E&US264@US64BusWen_E&NC96@US64/264 35.834385 -78.321147
I-25BLCas@US20/26Bus_W&US20BusCas@I-25BL_N&US26BusCas@I-25BL_N&US87BusCas@US20/26Bus_W 42.850139 -106.325254
GA316@GA10Lp(1)&US29@GA10Lp(1)&US78@GA10Lp(1)&GA8@GA10Lp(1)&GA10LpAth@1&GA10LpAth@US78 33.912799 -83.450979
US11AltRoa@I-581&I-581@4&US220AltRoa@US220_S&US220@I-581(4)&US460@I-581&VA116@I-581(4) 37.282214 -79.938498
US15BusWar@US211/29Bus_S&US29BusWar@US15/211Bus&US211BusWar@US211_End&US211@US15/29Bus 38.727857 -77.790831
US191BusLew@US191_S&US87@US87Bus_N&US191@US87/200&MT200@MT200Bus_W&MT200BusLew@MT200_W 47.055534 -109.476947
I-90BLShe@US87_S&US14BusShe@US87_S&US87BusShe@US87_S&US87She@US14Bus/87Bus&WY334@US87 44.773682 -106.942532
US17TrkOrl@US17_S&US17@US17/92Trk_S&US92TrkOrl@US92_W&US92@US17/92Trk_S&FL423@US17/92 28.605946 -81.365215
US1BusWay@US84_W&US23BusWay@US84_W&US84@US1/23Bus_S&GA4BusWay@US84_W&GA38@US1/23Bus_S 31.212095 -82.355849
US1BusWay@US84_E&US23BusWay@US84_E&US84@US1/23Bus_N&GA4BusWay@US84_E&GA38@US1/23Bus_N 31.213923 -82.353475
US70AltGib@US79/70AltByp&US70AltBypHum@US70Alt_W&US79BypHum@US79_S&US79@US79/70AltByp 35.813924 -88.933554
US60BusCli@US60Trk/188&US60TrkCli@US60_E&US220BusCli@US60Trk/188&VA188@US60Bus/220Bus 37.81609 -79.824761
A1@CHE/FRA&E21@FRA/CHE&E25@CHE/FRA&E25MtB@CHE/FRA&E62@FRA/CHE&E62@FRA/CHE&A41@FRA/CHE 46.148621 6.09561
I-2@176&I-69E@26B&US83BusMcA@US83_S&US77@I-69E(26B)&US83@I-69E(26B)&TXSpr206@I-2/69E 26.189545 -97.717697
IndNatTpk@1&US70BusHug@US70_W&US70@US70Bus_W&US271BusHug@US271_N&US271@US271BusHug_N 34.025757 -95.547709
I-90BLStu@I-90_W&I-90@30&US14AltDea@US14_E&US14@I-90(30)&SD34@I-90(30)&SD79@I-90(30) 44.41867 -103.531852
I-35BLCle@I-35_N&I-35@194&US18BusMas@US18_W&US18@I-35(194)&IA27@I-35(194)&IA122@I-35 43.146214 -93.355469
E5@27(AP8)&GI20@AP1/AP8_W&AP1@27(AP8)&AP8@27&E70@27(AP8)&E80@27(AP8)&GI20x@AP1/AP8_W 43.28928 -2.036752
E5@12(AP8)&GI20@AP1/AP8_E&AP1@12(AP8)&AP8@12&E70@12(AP8)&E80@12(AP8)&GI20x@AP1/AP8_E 43.304468 -1.89483
E61@SVN/ITA&E61@SVN/ITA&E70@ITA/SVN&E70@ITA/SVN&RA14@ITA/SVN&A3@SVN/ITA&SS58@ITA/SVN 45.700028 13.835757
VaajMoo@SulTie&E63@SulTie&Vt4@SulTie&Vt9@SulTie&Vt13@SulTie_N&Vt23@SulTie&E75@SulTie 62.241251 25.851302
US129AltMac/US129BusHaw/US129_S/US341/GA11_S/GA11BusHaw/GA26/GA27/GA112/GA230/GA257 32.28207 -83.466863
I-30@139B&I-630@1&US65@I-30(139B)&US67@I-30(139B)&US70@I-30(139B)&US167@I-30(139B) 34.733924 -92.261853
I-74FutLau@184&US15BusLau@FI-74&US74@FI-74(184)&US401BusLau@FI-74&US501@FI-74(184) 34.755479 -79.475119
I-26FutAsh@23&US19@US25_S&US19BusWea@US19_S&US23@US25_S&US25@US19/23_S&US70@US25_E 35.644848 -82.581268
E5@A10/A630&E5Bor@A10/A630&E70@1(A630)&A10@A630&A630@A10/N230&N230@1&E606@A10/A630 44.88461 -0.511808
US20BusCas@US20_W&US20@US20/26Bus&US26BusCas@US26_W&US26@US20/26Bus&WY257@US20/26 42.860764 -106.400603
US21ConCol@US21&US21@US21/76Con&US76ConCol@US21&US176@US21/76Con&US321@US21/76Con 33.989611 -81.041246
US521TrkCam@US521_N&US601TrkCam@US601_N&US521@US521/601Trk_N&US601@US521/601Trk_N 34.290499 -80.611326
US521BusKer@US521_S&US601BusKer@US601_S&US521@US521/601Bus_S&US601@US521/601Bus_S 34.541646 -80.582243
US15BusLau@US74Bus&US74BusLau@US15/401Bus&US401BusLau@US74Bus&US501BusLau@US74Bus 34.77412 -79.46277
US127BypDan@US150/150Byp&US150BypDan@US150_W&KY52@US127/150Byp&US150@US127/150Byp 37.648558 -84.803075
US211BusWas@US211_W&US522BusWas@US522_S&US211@US211/522Bus_S&US522@US211/522Bus_S 38.698356 -78.163304
US211BusWas@US211_E&US522BusWas@US522_N&US211@US211/522Bus_N&US522@US211/522Bus_N 38.711718 -78.152715
US191BusLew@US87_S&US87@US87Bus_S&MT200@MT200Bus_E&MT200BusLew@MT200_E&SR238@US87 47.066892 -109.424021
</pre>

**Comment by yakra** [comment]
`if you save as a .wpl file`
lat & lng are just bare numbers, not part of an OSM URL.

**Comment by yakra** [comment]
HutRivPkwy@1&I-95@6B&I-278@I-95&I-295@I-95&I-678@I-95

**Comment by yakra** [comment]
ME11@I-95(109)&ME17@I-95&ME100@I-95(109)&US202@I-95(109)

**Comment by yakra** [comment]
TX21@US183/130&TX130@461&US183@TX21

**Comment by yakra** [comment]
not a failsafe label, but:
I-95/US1/US9/US46/I-95/US1/US9@NJ/NY

**Comment by yakra** [comment]
Don't know if this is a duplicate of an example already covered, but:
ME113Cha@NH113B_N&NH113B@ME113_N

**Comment by yakra** [comment]
3+ intersection: N493@RingSpi&RingSpi@N493_E&RingSpi@N493_W -> N493_W/RingSpi/RingSpi

**Comment by jteresco** [comment]
I-93@12&MA3@12(93)&MA3APly@MA3_N&MA203@I-93&US1@I-93(12)

**Comment by yakra** [comment]
I-35Gai@478&US77@I-35(478)

**Comment by yakra** [comment]
TX36@US90Alt_E&US90AltHal@TX36_S

**Comment by yakra** [comment]
I-90@24&I-93@20&MA3@20(93)&US1@I-93(20)
Waypoint 1363

**Comment by yakra** [comment]
https://github.com/TravelMapping/DataProcessing/issues/90#issuecomment-464908517
> not a failsafe label, but:
`I-95/US1/US9/US46/I-95/US1/US9@NJ/NY`
**Edit:** similarly,
`E58/E81/E81/M26@UKR/ROU`

This happens because:
https://github.com/TravelMapping/DataProcessing/blob/8528fe9cbb5a9fa5260a2dc4a1e7042981a2d444/siteupdate/python-teresco/siteupdate.py#L465-L466
https://github.com/TravelMapping/DataProcessing/blob/88d12c5b0ac5a16c235354d383b04b10004c4322/siteupdate/cplusplus/classes/Waypoint/canonical_waypoint_name.cpp#L38-L39
By the time we get to the 2nd `I-95` or the 2nd `E81`, **routes** will not equal any one list_entry_name because it already contains more than one.

## Potential solution:
Replace **routes** with...
* a list, checking for membership as new pointname matches are discovered; or
* a set, re-sorting contents before converting to string

...whichever is faster *(list, maybe?)* and create a string from its contents when all pointnames match.

---

As an aside, `break`ing from the for loop when a pointname fails to match may save us a bit of time.

**Comment by yakra** [comment]
Made some temporary label changes as I start to text things out.
```DIFF
134818c134818
< Keep failsafe: ME24BusBru@US201_N&US201@ME24Bus_N
---
> Straightforward intersection: ME24BusBru@US201_N&US201@ME24BusBru_N -> ME24BusBru_N/US201_N
206730c206730
< Keep failsafe: US54BusElP@US54_E&US54ElP@US54Bus_E
---
> Keep failsafe: US54BusElP@US54_E&US54ElP@US54BusElP_E
```
The first result was expected; the second came as a surprise. It'll be interesting to get into the code and see what makes things tick.

**Comment by yakra** [comment]
## work in progress
**for:**
`Straightforward intersection: US54BusElP@US54_E&US54ElP@US54Bus_E -> US54BusElP_E/US54ElP_E`
`Straightforward intersection: ME24BusBru@US201_N&US201@ME24Bus_N -> ME24BusBru_N/US201_N`
```C++
if (ap_coloc.size() == 2)
{	//FIXME: there are more efficient ways to do this, but this will work for now as a proof of concept.
	std::string w0_list_entry = ap_coloc[0]->route->list_entry_name();
	std::string w1_list_entry = ap_coloc[1]->route->list_entry_name();
	std::string w0_no_abbrev  = ap_coloc[0]->route->name_no_abbrev();
	std::string w1_no_abbrev  = ap_coloc[1]->route->name_no_abbrev();
	std::string w0_label = ap_coloc[0]->label;
	std::string w1_label = ap_coloc[1]->label;
	if (	(w1_label == w0_list_entry or w1_label == w0_no_abbrev or w1_label.find(w0_list_entry + '_') == 0 or w1_label.find(w0_no_abbrev + '_') == 0)
	     &&	(w0_label == w1_list_entry or w0_label == w1_no_abbrev or w0_label.find(w1_list_entry + '_') == 0 or w0_label.find(w1_no_abbrev + '_') == 0)
	   )
	   {	const char *u0 = strrchr(w0_label.data(), '_');
		const char *u1 = strrchr(w1_label.data(), '_');
		std::string newname = (w0_list_entry + (u1 ? u1 : "")) + "/" + (w1_list_entry + (u0 ? u0 : ""));
		log.push_back("Straightforward intersection: " + name + " -> " + newname);
		return newname;
	   }
}
```
It increases processing time just a smidgen (~0.3s), not as bad as I'd anticipated.
As noted, there may be a more efficient solution; I'm hoping for something faster than what we started out with.
Leaving this here for reference for when I get started on the Python version.
We're now catching 5511 cases that previously were not detected until the Exit/Intersection check.

Some labels look a bit awkward, for example:
```
Straightforward intersection: L23Str@L38_N&L38Erk@L23 -> L23Str/L38Erk_N
Straightforward intersection: L23Str@L38_S&L38Fre@L23 -> L23Str/L38Fre_S
Straightforward intersection: E39Ber@Fv552_N&Fv552Ulv@E39 -> E39Ber/Fv552Ulv_N
Straightforward intersection: E39Ber@Fv552_S&Fv552Oso@E39 -> E39Ber/Fv552Oso_S
```

I'll play around with this some more and see what I can come up with.

---
Shell script to view only the new cases that started out as Keep failsafe:
```sh
labels=`diff logs/waypointsimplification.old logs/waypointsimplification.i3 | grep 'failsafe:' | cut -f4 -d' '`
for label in $labels; do diff logs/waypointsimplification.old logs/waypointsimplification.i3 | grep "$label"; done | less
```

**Comment by yakra** [comment]
> Some labels look a bit awkward, for example:
> ```
> Straightforward intersection: L23Str@L38_N&L38Erk@L23 -> L23Str/L38Erk_N
> Straightforward intersection: L23Str@L38_S&L38Fre@L23 -> L23Str/L38Fre_S
> Straightforward intersection: E39Ber@Fv552_N&Fv552Ulv@E39 -> E39Ber/Fv552Ulv_N
> Straightforward intersection: E39Ber@Fv552_S&Fv552Oso@E39 -> E39Ber/Fv552Oso_S
> ```
>
> I'll play around with this some more and see what I can come up with.
> **Edit:** I saw some _A _B _C suffixes, in... where was it, Quebec? Check these out.

After some further consideration, it looks like these should be left as-is. The example that clarified it for me:
```DIFF
> Straightforward intersection: PR10Pon@PR503_A&PR503Pon@PR10_S -> PR10Pon_S/PR503Pon_A
> Straightforward intersection: PR10Pon@PR503_C&PR503@PR10 -> PR10Pon/PR503_C
> Straightforward intersection: PR10Pon@PR503_B&PR503Pon@PR10_N -> PR10Pon_N/PR503Pon_B
```
Still a bit awkward at first blush, but it makes more sense than the alternative of swapping the suffixes around to the other routes.

> Look for city suffixes too.
```DIFF
> Straightforward intersection: MA2ACam@US3_N&US3@MA2A_Arl -> MA2ACam_Arl/US3_N
```
...makes it clearer still. [Opened a PR](https://github.com/TravelMapping/HighwayData/pull/3332) to change the label in US3 from `MA2A_Arl` to `MA2_W`. However, the `_Arl` would belong with the MA2 portion of the label in the same way the `_W` does.
We shouldn't have a `US3_W`, as US3 splits N.
We shouldn't have a `MA2_N`, as MA2 splits W.
*Yes! No swapping suffixes.*

---

We still will get some occasional odd-looking results:
```DIFF
< Keep failsafe: TN269Eag@US41Alt&US41AltNas@TN269_Eag
> Straightforward intersection: TN269Eag@US41Alt&US41AltNas@TN269_Eag -> TN269Eag_Eag/US41AltNas
< Keep failsafe: I-90BLMoo@US16_E&US16@I-90BL_Moo
> Straightforward intersection: I-90BLMoo@US16_E&US16@I-90BL_Moo -> I-90BLMoo_Moo/US16_E
< Keep failsafe: OR223Dal@OR223_Ric&OR223Ric@OR223_Dal
> Straightforward intersection: OR223Dal@OR223_Ric&OR223Ric@OR223_Dal -> OR223Dal_Dal/OR223Ric_Ric
```
This happens because I'm inserting the abbrevs and then appending the label's original _suffix.
I'll try going back to using just a `w1_label + "/" + w0_label` style as per @jteresco's original implementation. I wanna say that should be sufficient to create unambiguous labels.

**Comment by yakra** [comment]
> I wanna say that should be sufficient to create unambiguous labels.

Nope. Pure brainfart. We have 100 new cases of `Appended region`, from 3 of which we `Revert to simple`.
I want a more elegant solution; this will take some brainstorming.

(The shell script [above](https://github.com/TravelMapping/DataProcessing/issues/90#issuecomment-555724226) doesn't let us see the whole picture; it filters out new Straightforward intersections which used to be Exit/Intersection cases, and thus we don't see what happens to them farther down the line.)

**Comment by yakra** [comment]
### "Straightforward intersection" gets a mulligan.

Current code, for reference purposes. I'd still like to try some hand-tuned C-string based speed improvements.
```C++
if (ap_coloc.size() == 2)
{	//FIXME: there are more efficient ways to do this, but this will work for now as a proof of concept.
	std::string w0_list_entry = ap_coloc[0]->route->list_entry_name();
	std::string w1_list_entry = ap_coloc[1]->route->list_entry_name();
	std::string w0_no_abbrev  = ap_coloc[0]->route->name_no_abbrev();
	std::string w1_no_abbrev  = ap_coloc[1]->route->name_no_abbrev();
	std::string w0_label = ap_coloc[0]->label;
	std::string w1_label = ap_coloc[1]->label;
	if (	(w1_label == w0_list_entry or w1_label == w0_no_abbrev or w1_label.find(w0_list_entry + '_') == 0 or w1_label.find(w0_no_abbrev + '_') == 0)
	     &&	(w0_label == w1_list_entry or w0_label == w1_no_abbrev or w0_label.find(w1_list_entry + '_') == 0 or w0_label.find(w1_no_abbrev + '_') == 0)
	   )
	   {	std::string newname = w1_label+"/"+w0_label;
		// if this is taken, attempt to add in abbrevs if there's point in doing so
		if ( (ap_coloc[0]->route->abbrev.size() || ap_coloc[1]->route->abbrev.size()) && vertex_names.find(newname) != vertex_names.end() )
		{	const char *u0 = strrchr(w0_label.data(), '_');
			const char *u1 = strrchr(w1_label.data(), '_');
			std::string newname = (w0_list_entry + (u1 ? u1 : "")) + "/" + (w1_list_entry + (u0 ? u0 : ""));
			log.push_back("Straightforward intersection: " + name + " -> " + newname + " (" + w1_label+"/"+w0_label + " already taken)");
			return newname;
		}
		log.push_back("Straightforward intersection: " + name + " -> " + newname);
		return newname;
	   }
}
```

Initially create a shortened label without including abbrevs.
If that's taken, add the abbrev(s) if available.
If not available, keep the initial *abbrev*-less one, falling through as before to append a region code.
For example:
```
waypointsimplification.i5:49133:Straightforward intersection: A1Ara@A6&A6@A1 -> A1/A6
--
waypointsimplification.i5:208241:Straightforward intersection: A1@A6&A6@A1 -> A1/A6
waypointsimplification.i5:208242:Appended region: A1/A6|AUS-QLD
--
waypointsimplification.i5:209650:Straightforward intersection: A1Syd@A6&A6@A1 -> A1Syd/A6 (A1/A6 already taken)
```

---

### a trail of breadcrumbs...

To examine the new appended labels in detail...
```sh
labels=`diff waypointsimplification.old waypointsimplification.i5 | egrep 'Append' | cut -f4 -d' ' | cut -f1 -d'|'`
for CanLabel in $labels; do
  part1=`echo $CanLabel | cut -f1 -d/`
  part2=`echo $CanLabel | cut -f2 -d/`
  variants="$part1[A-Za-z_]*/$part2[A-Za-z_]*"

  SimList=`egrep " $variants$" waypointsimplification.i5 | cut -f2 -d: | cut -f2 -d' '`
  SimLabels=""
  for sl in $SimList; do
    SimLabels="$SimLabels$sl |$sl$|"
  done

  clear
  egrep -Hn -C 0 "$SimLabels $variants[ \|]| $variants$" waypointsimplification.old
  echo -e "\n\n\n================================================================================\n\n\n"
  egrep -Hn -C 0 "$SimLabels $variants[ \|]| $variants$" waypointsimplification.i5
  echo -en 'Press Enter '
  read OK
done
```
To find the region for the canonical version of each of these labels:
`for label in $labels; do grep "^$label " *region-collapsed.tmg; done`

We avoid appending the region to 74 waypoints, as these get abbrevs to avoid a collision now:
`grep " already taken)" waypointsimplification.i5 | less`

---
We still have **26** new cases of **Appended region**.
* Of these, 2 are for labels where we'd be appending a region anyway:
  - 2 plain abbrevless `L161@L162&L162@L161` labels to simplify, in DEU-ST & DEU-BW.
  - 3 plain abbrevless `A1@A3&A3@A1` labels to simplify, in MAR, AUS-QLD & DEU-BW.

Otherwise...
* 1 case is a multiplex split, and can be avoided by [adding directional suffixes to the underlying waypoints](http://forum.travelmapping.net/index.php?topic=3334.0), resulting in `L1137_W/L1141_S`. [Already fixed in the master branch](https://github.com/TravelMapping/HighwayData/pull/3338).
* The other 25? These all happen when points with abbrevs get processed *first*, and then at a 2nd location, we find points with no abbrevs to add. In theory, these can be avoided too.
  - It's possible to force an earlier vertex to look for a new name, but that's too much labor for too little gain. Better yet...
  - The simpler & more useful option would be to change how colocation lists are sorted, ensuring same-designation routes without abbrevs are sorted before those with.
    - Gotta learn how to do this in Python.
    - Would mess up existing `VISIBLE_HIDDEN_COLOC` & `HIDDEN_JUNCTION` FPs, but these could be replaced automatically.
* Only two cases of two like-named labels in the same region:
  - `B412/L82` (in DEU-RP) and `B412/L82|DEU-RP` -- less useful than the old `B412Bur@L82&L82Sin@B412` and `B412/L82`.
  - `EO8/EO8A` (in GRC) and `EO8/EO8A|GRC` -- less useful than the old `EO8@EO8A&EO8AEle@EO8` and `EO8/EO8A`.
* So yeah. Just two really suboptimal labels IMO, that's not bad. And they're still shorter and easier to read. Along with everything else getting shortened. Here's the breakdown, of our **9087** new "Straightforward intersection" cases:
  - **3573** are newly shortened, and used to be *Keep failsafe* before we started.
  - **5511** used to be *Exit/Intersection* cases, and are now shortened earlier in the process. Of these:
    - **5500** are shorter than the old *Exit/Intersection* version.
    - **11** are identical to the old *Exit/Intersection* version.
    - Nothing longer, or different with the same length.
  - This leaves us with **3** remaining rather curious cases. More on these in my next post...

But first, the script I used to examine the old *Exit/Intersection* cases:
```sh
SimLabels=`diff waypointsimplification.old waypointsimplification.i5 | egrep '^< Exit/I' | cut -f3 -d' '`
echo "Finding old & new canonical labels for former Exit/Intersection cases. Please wait..."
CanLabels=`for label in $SimLabels; do grep " $label " waypointsimplification.i5 | cut -f3 -d' '; done`
for CanLabel in $CanLabels; do
  OldShort=`grep $CanLabel waypointsimplification.old | cut -f '2' -d'>' | cut -f2 -d' '`
  NewShort=`grep $CanLabel waypointsimplification.i5  | cut -f '2' -d'>' | cut -f2 -d' '`
  echo -en "$CanLabel\t$OldShort\t$NewShort\t"
  if [ $OldShort = $NewShort ]; then
    echo Identical
  else
    if   [ `echo $NewShort | wc -c` < `echo $OldShort | wc -c` ]; then
      echo Shorter
    elif [ `echo $NewShort | wc -c` > `echo $OldShort | wc -c` ]; then
      echo Longer
    else echo NewSameLength
    fi
  fi
done
```

**Comment by yakra** [comment]
### Our 3 remaining curious cases:
~~~
yakra@BiggaTomato:~/TravelMapping/yakra/DataProcessing/siteupdate/cplusplus/logs$ diff waypointsimplification.old waypointsimplification.i5 | egrep '^< Straightforward concurrency'
< Straightforward concurrency: L3238Her@L3238&L3238Mar@L3238 -> L3238Her/L3238Mar@L3238
< Straightforward concurrency: B279@B279&B279Hol@B279 -> B279/B279Hol@B279
< Straightforward concurrency: A329@A329_E&A329Rea@A329_E -> A329/A329Rea@A329_E
yakra@BiggaTomato:~/TravelMapping/yakra/DataProcessing/siteupdate/cplusplus/logs$ egrep 'L3238Her@L3238&L3238Mar@L3238|B279@B279&B279Hol@B279|A329@A329_E&A329Rea@A329_E' waypointsimplification.i5
Straightforward intersection: L3238Her@L3238&L3238Mar@L3238 -> L3238/L3238
Straightforward intersection: B279@B279&B279Hol@B279 -> B279/B279
Straightforward intersection: A329@A329_E&A329Rea@A329_E -> A329_E/A329_E
~~~
**ORIGINAL** <br> *simple,* <br> *"failsafe"* | **OLD** <br> *Straightforward concurrency* | **NEW** <br> *Straightforward intersection*
-- | -- | --
L3238Her@L3238&L3238Mar@L3238 | L3238Her/L3238Mar@L3238 | L3238/L3238
B279@B279&B279Hol@B279 | B279/B279Hol@B279 | B279/B279
A329@A329_E&A329Rea@A329_E | A329/A329Rea@A329_E | A329_E/A329_E

As we can see, the new labels are less than clear.
Even in an alternate universe where the underlying waypoint labels contained *abbrev*s, we'd still get the same new shortened labels, because *abbrev*s are thrown out unless needed for disambiguating.

It makes sense to add in a check for matching `name_no_abbrev()`s and add in the *abbrev*s in these cases.
*So what's the fastest way to do it, mumble mumble... Probably that way...*

**Comment by yakra** [comment]
And now:
```
yakra@BiggaTomato:~/TravelMapping/yakra/DataProcessing/siteupdate/cplusplus/logs$ egrep 'L3238Her@L3238&L3238Mar@L3238|B279@B279&B279Hol@B279|A329@A329_E&A329Rea@A329_E' waypointsimplification.i6
Straightforward intersection: L3238Her@L3238&L3238Mar@L3238 -> L3238Her/L3238Mar
Straightforward intersection: B279@B279&B279Hol@B279 -> B279/B279Hol
Straightforward intersection: A329@A329_E&A329Rea@A329_E -> A329_E/A329Rea_E
```
The other diffs affected by this change:
```sh
yakra@BiggaTomato:~/TravelMapping/yakra/DataProcessing/siteupdate/cplusplus/logs$ echo "simple/failsafe | last version | this version"; echo '| -- | -- | -- |'; paste -d'|' <(diff waypointsimplification.i5 waypointsimplification.i6 | grep '^<' | sed -e 's~< Straightforward intersection: ~~' -e 's~\->~|~' -e 's~\(^.*$\)~\1 ~') <(diff waypointsimplification.i5 waypointsimplification.i6 | grep '^>' | cut -f3 -d'>')
```
simple/failsafe | last version | this version
| -- | -- | -- |
Fv835Als@Fv835Lei&Fv835Lei@Fv835 | Fv835/Fv835Lei | Fv835Als/Fv835Lei
Fv64And@Fv64&Fv64Isf@Fv64And | Fv64And/Fv64 | Fv64And/Fv64Isf
Fv61Har@Fv61Uls&Fv61Uls@Fv61 | Fv61/Fv61Uls | Fv61Har/Fv61Uls
Fv120Gar@Fv120&Fv120Tei@Fv120Gar | Fv120Gar/Fv120 | Fv120Gar/Fv120Tei
Fv552Oso@Fv552Ulv&Fv552Ulv@Fv552 | Fv552/Fv552Ulv | Fv552Oso/Fv552Ulv
Fv544Hei@Fv544Lei&Fv544Lei@Fv544 | Fv544/Fv544Lei | Fv544Hei/Fv544Lei
L193Els@L193Tec&L193Tec@L193 | L193/L193Tec | L193Els/L193Tec
L47Alt@L47&L47Ren@L47Alt | L47Alt/L47 | L47Alt/L47Ren
B475Lad@B475Sae&B475Sae@B475 | B475/B475Sae | B475Lad/B475Sae
L3238Her@L3238&L3238Mar@L3238 | L3238/L3238 | L3238Her/L3238Mar
L3251Her@L3251_E&L3251Wom@L3251Her_S | L3251Her_S/L3251_E | L3251Her_S/L3251Wom_E
L3251Her@L3251_W&L3251Wom@L3251Her_N | L3251Her_N/L3251_W | L3251Her_N/L3251Wom_W
L3225Gro@L3225Hes&L3225Hes@L3225 | L3225/L3225Hes | L3225Gro/L3225Hes
L609Dor@L609Oes&L609Oes@L609 | L609/L609Oes | L609Dor/L609Oes
L183Bon@L183Bor&L183Bor@L183 | L183/L183Bor | L183Bon/L183Bor
B9Dor@B9Kol&B9Kol@B9 | B9/B9Kol | B9Dor/B9Kol
L241Lam@L241Wei&L241Wei@L241 | L241/L241Wei | L241Lam/L241Wei
S111Ams@S111Ove&S111Ove@S111 | S111/S111Ove | S111Ams/S111Ove
B2Hof@B2Zed&B2Zed@B2 | B2/B2Zed | B2Hof/B2Zed
B15Alt@B15&B15Tir@B15Alt | B15Alt/B15 | B15Alt/B15Tir
B279@B279&B279Hol@B279 | B279/B279 | B279/B279Hol
B327Kob@B327Obe&B327Obe@B327 | B327/B327Obe | B327Kob/B327Obe
B33Kon@B33&B33Rad@B33Kon | B33Kon/B33 | B33Kon/B33Rad
SS336DirMal@SS336DirVan&SS336DirVan@SS336Dir | SS336Dir/SS336DirVan | SS336DirMal/SS336DirVan
D565Mur@D565_N&D565Tek@D565Mur_N | D565Mur_N/D565_N | D565Mur_N/D565Tek_N
D565Mur@D565_S&D565Tek@D565Mur_S | D565Mur_S/D565_S | D565Mur_S/D565Tek_S
D330Mug@D330Yes_E&D330Yes@D330_E | D330_E/D330Yes_E | D330Mug_E/D330Yes_E
D555Akh@D555&D555Sal@D555Akh | D555Akh/D555 | D555Akh/D555Sal
SS713DirBro@SS713Dir&SS713DirSSB@SS713DirBro | SS713DirBro/SS713Dir | SS713DirBro/SS713DirSSB
A167Bla@A167&A167Nun@A167Bla | A167Bla/A167 | A167Bla/A167Nun
A1036@A1036_E&A1036Yor@A1036_S | A1036_S/A1036_E | A1036_S/A1036Yor_E
A329@A329_E&A329Rea@A329_E | A329_E/A329_E | A329_E/A329Rea_E
PA420AltTrkPro@PA420AltTrk_Rid&PA420AltTrkRid@PA420AltTrk_Pro | PA420AltTrk_Pro/PA420AltTrk_Rid | PA420AltTrkPro_Pro/PA420AltTrkRid_Rid
OR223@OR223_Dal&OR223Dal@OR223 | OR223/OR223_Dal | OR223/OR223Dal_Dal
OR223Dal@OR223_Ric&OR223Ric@OR223_Dal | OR223_Dal/OR223_Ric | OR223Dal_Dal/OR223Ric_Ric

Current state of straightforward_intersection.cpp, minus the opening comments:
```C++
if (ap_coloc.size() == 2)
{	//FIXME: there are more efficient ways to do this, but this will work for now as a proof of concept.
	std::string w0_list_entry = ap_coloc[0]->route->list_entry_name();
	std::string w1_list_entry = ap_coloc[1]->route->list_entry_name();
	std::string w0_no_abbrev  = ap_coloc[0]->route->name_no_abbrev();
	std::string w1_no_abbrev  = ap_coloc[1]->route->name_no_abbrev();
	std::string w0_label = ap_coloc[0]->label;
	std::string w1_label = ap_coloc[1]->label;
	if (	(w1_label == w0_list_entry or w1_label == w0_no_abbrev or w1_label.find(w0_list_entry + '_') == 0 or w1_label.find(w0_no_abbrev + '_') == 0)
	     &&	(w0_label == w1_list_entry or w0_label == w1_no_abbrev or w0_label.find(w1_list_entry + '_') == 0 or w0_label.find(w1_no_abbrev + '_') == 0)
	   ) {	std::string newname = w1_label+"/"+w0_label;
		// if this is taken or if name_no_abbrev()s match, attempt to add in abbrevs if there's point in doing so
		bool taken = vertex_names.find(newname) != vertex_names.end();
		if ( 	(ap_coloc[0]->route->abbrev.size() || ap_coloc[1]->route->abbrev.size())
		     && (taken || w0_no_abbrev == w1_no_abbrev)
		   ) {	const char *u0 = strrchr(w0_label.data(), '_');
			const char *u1 = strrchr(w1_label.data(), '_');
			std::string newname = (w0_list_entry + (u1 ? u1 : "")) + "/" + (w1_list_entry + (u0 ? u0 : ""));
			std::string message = "Straightforward intersection: " + name + " -> " + newname;
			if (taken) message += " (" + w1_label+"/"+w0_label + " already taken)";
			log.push_back(message);
			return newname;
		     }
		log.push_back("Straightforward intersection: " + name + " -> " + newname);
		return newname;
	     }
}
```

## ToDo:
* See how execution time is doing
  - Probably no performance hit from reading `ap_coloc[0]->label` directly rather than storing it as `w0_label`, right?
  - Work on C-string based speedup
* Longer term, implement a new sorting method for colocation lists

**Comment by yakra** [comment]
```C++
const char *u0 = strrchr(w0_label.data(), '_');
const char *u1 = strrchr(w1_label.data(), '_');
```
Less efficient than `strchr`; we already need to scan forward from from the **char\*** location to find our terminating `\0`. Waypoint labels should only have one underscore anyway; if not, this is the more robust way to handle `LABEL_UNDERSCORES` cases.

**Comment by yakra** [comment]
# i7 & i8
**i7:** use `Waypoint::label_references_single_route(Route *r)`
**i8:** comment out final check for underscore or null zero

---
>   - Work on C-string based speedup

OK, so this ended up using the C++ string class; it's just more convenient. :P
No noticeable change in speed; keeping it because in addition to a shorter `straightforward_intersection.cpp`, we gain a new `Waypoint::label_references_single_route(Route *r)` function that may prove useful in future waypoint simplification or datacheck enhancements.

```C++
bool Waypoint::label_references_single_route(Route *r)
{	std::string no_abbrev = r->name_no_abbrev();
	if (label.substr(0, no_abbrev.size()) != no_abbrev)
		return 0;
	if (label[no_abbrev.size()] == 0 || label[no_abbrev.size()] == '_')
		return 1;
	if (label.substr(no_abbrev.size(), r->abbrev.size()) != r->abbrev)
		return 0;
	if (label[no_abbrev.size() + r->abbrev.size()] == 0 || label[no_abbrev.size() + r->abbrev.size()] == '_')	// If I comment out this line along with...
		return 1;
	else	return 0;												// ...this one, then the results change.
}
```

```C++
if (ap_coloc.size() == 2)
  if (ap_coloc[1]->label_references_route(ap_coloc[0]->route) && ap_coloc[0]->label_references_route(ap_coloc[1]->route))
  {	std::string newname = ap_coloc[1]->label+"/"+ap_coloc[0]->label;
	// if this is taken or if name_no_abbrev()s match, attempt to add in abbrevs if there's point in doing so
	bool taken = vertex_names.find(newname) != vertex_names.end();
	if ( 	(ap_coloc[0]->route->abbrev.size() || ap_coloc[1]->route->abbrev.size())
	     && (taken || ap_coloc[0]->route->name_no_abbrev() == ap_coloc[1]->route->name_no_abbrev())
	   ) {	const char *u0 = strchr(ap_coloc[0]->label.data(), '_');
		const char *u1 = strchr(ap_coloc[1]->label.data(), '_');
		std::string newname = (ap_coloc[0]->route->list_entry_name() + (u1 ? u1 : "")) + "/" + (ap_coloc[1]->route->list_entry_name() + (u0 ? u0 : ""));
		std::string message = "Straightforward intersection: " + name + " -> " + newname;
		if (taken) message += " (" + ap_coloc[1]->label+"/"+ap_coloc[0]->label + " already taken)";
		log.push_back(message);
		return newname;
	     }
	log.push_back("Straightforward intersection: " + name + " -> " + newname);
	return newname;
  }
```

Commenting out the two lines noted above in the new function changes how we react to an unexpected abbrev, letter suffix, or slashed designation giving us 228 diffs. I'm still in the process of looking through them, and the results are... interesting. I see evidence of cases where editing went awry; some potential for future datachecks. In almost all cases, Jim's existing labels (which we get with the two lines of code left in place) are objectively better.
I'll keep looking at the data, investigate possibilities for new datachecks, and decide how to proceed.

**Comment by yakra** [comment]
i1 thru 19 testing thus far has used
HighwayData @ 453868ffca54e08678a40a4e8f432e011820d2e4
UserData @ 3ac73953ace43690ec1e1ede01f73257a04977c5

**Comment by yakra** [comment]
> **The simpler & more useful option would be to change how colocation lists are sorted,** ensuring same-designation routes without abbrevs are sorted before those with. 

That won't work. We're iterating through waypoints supplied from the quadtree, meaning they're ordered by quadtree node, which has no relation to abbrevs, strings, anything in the CSVs.

Instead:
* Iterate through highway_systems
  * Sort route_list by root. If we don't, NS4 will be sorted after NS4Xyz, etc.
  * Iterate through that.
    * Iterate through point_list.

**Comment by yakra** [comment]
> * Iterate through highway_systems
>   * Sort route_list by root. If we don't, NS4 will be sorted after NS4Xyz, etc.
>   * Iterate through that.

Still won't work. We process `deubbl` before `deunwl`, meaning we process `deubb.l023str` before `deunw.l023`.

Try making one master list of all routes, and sorting it by `list_entry_name()`.

**Comment by yakra** [comment]
25 of [**26** new cases of **Appended region**](https://github.com/TravelMapping/DataProcessing/issues/90#issuecomment-557266834) are fixed. 1 still suboptimal:
```
waypointsimplification.i10:105760:Straightforward intersection: EO8@EO8A&EO8AEle@EO8 -> EO8/EO8A
waypointsimplification.i10:105761:Exit/Intersection: EO8@EO8a_E&EO8ACor@EO8 -> EO8/EO8ACor
waypointsimplification.i10:105762:Exit/Intersection: EO8@EO8a_W&EO8A@EO8 -> EO8/EO8A
waypointsimplification.i10:105763:Appended region: EO8/EO8A|GRC
```

Sort by bool: Give priority to points with no abbrevs on any colocated route. *Simple!*

**Comment by yakra** [comment]
```C++
bool waypoint_simplification_sort(Waypoint *w1, Waypoint *w2)
{	if (	w1->colocated
	     && w1->colocated->size() == 2
	     && w1->colocated->front()->route->abbrev.empty()
	     && w1->colocated->back()->route->abbrev.empty()
	   ) return 1;
	else return 0;
}
```
`EO8/EO8A|GRC` example fixed. The other 25 still behave as before.

Even better:
```
yakra@BiggaTomato:~/TravelMapping/yakra/DataProcessing/siteupdate/cplusplus/logs$ grep -c Appended waypointsimplification.old waypointsimplification.i7 waypointsimplification.i10 waypointsimplification.i11
waypointsimplification.old:6864
waypointsimplification.i7:6890
waypointsimplification.i10:6870
waypointsimplification.i11:6864
```

**Comment by yakra** [comment]
# Nope.
> && w1->colocated->size() == 2

I should not be checking the number of *total* colocated points, but only the number of *active/preview* colocated points.

**Comment by yakra** [comment]
Finally got the results I was looking for.

**Speed:**
Time spent `Creating unique names and vertices` has slightly more than doubled. Make up for it by:
* Storing & retrieving `ap_coloc` vectors.
  - Avoid computing again during `canonical_waypoint_name`.
  - Avoid multiple calls during `waypoint_simplification_sort` comparisons.
* Don't sort (and thus create `ap_coloc` vectors for) all waypoints & then iterate thru them.
  Instead, have the quadtree return only the points corresponding to each vertex, creating vectors for each, and iterate thru that subset.
  - Remove those checks from beginning of `HighwayGraph` ctor loop.

**Results:** Longer results were originally what -- Exit/Intersection? Try switching order...
**Idea:** Simplify suffixes when matching abbrev

**Comment by yakra** [comment]
> **Results:** Longer results were originally Exit/Intersection. Try switching order...

Derp, nope. The shorter results were also Exit/Intersection. :P
Having 38 labels averaging 1.13 chars longer is the price we pay for getting 8977 labels averaging 7.58 chars shorter. :)

>  **Idea:** Simplify suffixes that match abbrev

Rejected. Would only help 3 labels. Too much complication for too little gain.

---

Some stats: **9066** labels have changed compared to the existing label simplification...

* **8977** shorter labels, by 7.58 chars average.
  - **3** formerly `Straightforward Concurrency`, where the `name_no_abbrev` of both routes matched.
    - 6, 5 and 3 chars shorter
  - **3573** formerly `Keep failsafe`
    - ranging from 6 to 27 chars shorter, 14.73 on average
  - The rest were formerly `Exit/Intersection`...
  - **4997** cases, shorter by 3 chars, where abbrevs are discarded and  `name_no_abbrev`s are slashed together, either in the same or opposite order.
  - **404** cases, shorter by 1 char, that lose a city *abbrev* and gain an underscored directional or A/B/C suffix.
* **51** labels the same length. All were originally `Exit/Intersection`, with the .list names slashed together in the opposite order.
* **38** labels, originally `Exit/Intersection`, are longer:
  - **35** by one char, **2** by two chars. These lose a city *abbrev* and gain an underscored city suffix, which matches the *abbrev* about half the time.
  - **1** by four chars. This gains an underscored city suffix while keeping its *abbrev* because we're dealing with two different flavors of unbannered OR223.

**Comment by yakra** [comment]
**Waypoint::label_references_route**
```C++
	if (	label[no_abbrev.size() + r->abbrev.size()] == '/'
		// exclusion for MO/WI county routes
	&& !(	isupper(label[no_abbrev.size() + r->abbrev.size()+1])
	     && (	label[no_abbrev.size() + r->abbrev.size()+2] == 0
		  ||	(isupper(label[no_abbrev.size() + r->abbrev.size()+2]) && label[no_abbrev.size() + r->abbrev.size()+3] == 0)
		)
	    )	// end exclusion for MO/WI county routes
	   )	datacheckerrors->add(route, label, "", "", "UNEXPECTED_DESIGNATION", label.substr(no_abbrev.size()+r->abbrev.size()+1));
	return 0;
```
Removing the exclusion & commenting out what's left of the datacheck for now.

**Comment by yakra** [comment]
```SH
cd ~/TravelMapping/devel
( echo 'latitude;longitude;old label;new label'
  paste -d';' \
  <(diff <(head -n 588644 i0can.tmg) \
	 <(head -n 588644 i15can.tmg) | grep '^<' | cut -f3 -d' ') \
  <(diff <(head -n 588644 i0can.tmg) \
	 <(head -n 588644 i15can.tmg) | grep '^<' | cut -f4 -d' ') \
  <(diff <(head -n 588644 i0can.tmg) \
	 <(head -n 588644 i15can.tmg) | grep '^<' | cut -f2 -d' ') \
  <(diff <(head -n 588644 i0can.tmg) \
	 <(head -n 588644 i15can.tmg) | grep '^>' | cut -f2 -d' ') \
) > i15diff.csv
```
• replace 588644 with num vertices + 2
• replace i0can & i15can with whatever canonical TMGs I want to diff

**Comment by yakra** [comment]
I-95@182&I-395@1&ME15@I-95/395

**Comment by yakra** [comment]
CT8@39&CT222@US6/8&US6@CT8(39)
